// @linthtml/dom-utils/dist/src/dom_elements.js
import { NodeWithChildren } from "domhandler";
import { ElementType } from "domelementtype";
// TODO: Add new type or template for CharValue with non empty raw value (like attribute value)
export class CharValue {
    constructor(chars, loc, raw) {
        this.chars = chars;
        this.loc = loc;
        this.raw = raw;
        this.chars = chars;
        this.raw = raw;
        this.loc = loc;
    }
}
export class NodeAttribute {
    constructor(name, loc, index, equal = null, value = null) {
        this.name = name;
        this.loc = loc;
        this.index = index;
        this.equal = equal;
        this.value = value;
        this.type = "attribute";
        this.name = name;
        this.index = index;
        this.value = value;
        this.equal = equal;
        this.loc = loc;
    }
}
export class Range {
    constructor(start, end) {
        this.start = start;
        this.end = end;
        this.start = start;
        this.end = end;
    }
}
export class Position {
    constructor(line, column) {
        this.line = line;
        this.column = column;
        this.line = line;
        this.column = column;
    }
}
/** @hidden */
function ExtendedNode(Base) {
    return class TExtractor extends Base {
        constructor() {
            super(...arguments);
            this._close = undefined;
        }
        get loc() {
            return this._loc;
        }
        set loc(value) {
            this._loc = value;
        }
        get open() {
            return this._open;
        }
        set open(value) {
            this._open = value;
        }
        get close() {
            return this._close;
        }
        set close(value) {
            this._close = value;
        }
    };
}
// Keep in sync with domhandler types
// Change NodeWithChildren type for children
export class Node extends ExtendedNode(NodeWithChildren) {
    constructor() {
        super(...arguments);
        this.parent = null;
        // eslint-disable-next-line no-use-before-define
        this.children = []; // {Node[]}
    }
}
/**
 * An element within the DOM.
 */
export class Element extends Node {
    /**
     * @param name Name of the tag, eg. `div`, `span`.
     * @param attribs Object mapping attribute names to attribute values.
     * @param children Children of the node.
     */
    constructor(name, attributes, children = [], type = name === "script"
        ? ElementType.Script
        : name === "style"
            ? ElementType.Style
            : ElementType.Tag) {
        super(type, children);
        this.name = name;
        this.attributes = attributes;
    }
    // DOM Level 1 aliases
    get tagName() {
        return this.name;
    }
    set tagName(name) {
        this.name = name;
    }
}
/**
 * The root node of the document.
 */
export class Document extends Node {
    constructor(children) {
        super(ElementType.Root, children);
    }
}
